### PR TITLE
finishing chapter10

### PR DESCRIPTION
**Pull request by 29jinghong**

### CODE REVIEW

File: programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            
+    private class UniformElement(
+        ch: Char,
+        override val width: Int,
+        override val height: Int
+        ) extends Element {
+            private val line = ch.toString * width
+            def contents = Array.fill(height)(line)        
+        }
+        def elem(contents:  Array[String]): Element =
```
Comment (wanggang1) [inline programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala:43]: the indentations are wrong for the 3 `def elem`

File: programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        def elem(chr: Char, width: Int, height: Int): Element =
+            new UniformElement(chr, width, height)
+        def elem(line: String): Element =
+            new LineElement(line)
+}
+
+abstract class Element {
+    def contents: Array[String]
+    def width: Int =
+        if (height == 0) 0 else contents(0).length
```
Comment (wanggang1) [inline programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala:54]: a more functional way: `cotents.headOption.map(_.length).getOrElse(0)`

**Review by wanggang1 — APPROVED** [review]
LGTM

**Comment by wanggang1** [comment]
Related to Ch 11
Screen Shot 2024-01-07 at 9 06 56 AM

File: programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala (hunk 2 above)
Comment (29jinghong) [reply programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala:54]: just wrote that in chapter 11 pr

File: programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala (hunk 1 above)
Comment (29jinghong) [reply programming-in-scala/src/main/scala/org/programingscala/chapter10/Chapter10.scala:43]: fixed it in chapter 11 pr

**Pull request merged**